### ISSUE TITLE
Autoregressive generation?

### ISSUE THREAD

**Issue by jlian2** [kyutai-labs/moshi#176]
### Due diligence

- [X] I have done my due diligence in trying to find the answer myself.

### Topic

The PyTorch implementation

### Question

I am just confused if the current model is the autoregressive modeling. I checked the step function in lm.py. It seems every time we just take the current frame as input to predict the next frame. It seems the history tokens (t=0,..,T) are not taken as input. In that case, how does moshi capture long-term dependency>

`    @torch.no_grad()
    def step(self, input_tokens: torch.Tensor) -> torch.Tensor | None:
        
        ##we should also check if text token is the input
        state = self._streaming_state
        # print(f"state cache shape is {state.cache.shape}") #[1, 17, 3]
        if state is None:
            raise RuntimeError(
                "You should wrap those calls with a `with lm_gen.streaming(): ...`."
            )
        lm_model = self.lm_model

        assert input_tokens.dim() == 3, "Shape should be [B, K, T]."
        B, Ki, S = input_tokens.shape
        assert S == 1, "Only support being given steps one by one."
        needed_tokens = lm_model.num_codebooks - lm_model.dep_q - 1
        assert (
            Ki == needed_tokens
        ), f"We expect {needed_tokens} tokens from the user stream, got {Ki}."`

**Comment by jlian2** [comment]
Understand now.
